BehaviorDelegate fails for suspend functions

retrofit-mock's `BehaviorDelegate` throws an exception when attempting to return for a suspend function.

For example
```kotlin
interface SomeApi {
  @GET("foo")
  suspend fun getFoo(): Foo
}
```
will throw when attempting to use:
```kotlin
val delegate: BehaviorDelegate<SomeApi> = mockRetrofit.create(SomeApi::class.java)
val foo: Foo = delegate.returningResponse(Foo()).getFoo()
```

I'd submit a failing test for this, but it would involve me pulling in Kotlin into the retrofit-mock module which I'm not sure is desirable?

Exception thrown looks like:
```
java.lang.IllegalArgumentException: Could not locate call adapter for class java.lang.Object.
 Tried:
  * retrofit2.CompletableFutureCallAdapterFactory
  * retrofit2.DefaultCallAdapterFactory
   at retrofit2.Retrofit.nextCallAdapter(Retrofit.java:241)
   at retrofit2.Retrofit.callAdapter(Retrofit.java:205)
   at retrofit2.mock.BehaviorDelegate$1.invoke(BehaviorDelegate.java:64)
   at java.lang.reflect.Proxy.invoke(Proxy.java:1006)
   at $Proxy0.getGames(Unknown Source)
   at au.com.gridstone.debugdrawer.sampleapp.MockGamesApi.getGames(MockGamesApi.kt:17)
   at au.com.gridstone.debugdrawer.sampleapp.GamesViewModel$refresh$1.invokeSuspend(GamesViewModel.kt:43)
   at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
   at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:238)
   at android.os.Handler.handleCallback(Handler.java:883)
   at android.os.Handler.dispatchMessage(Handler.java:100)
   at android.os.Looper.loop(Looper.java:214)
   at android.app.ActivityThread.main(ActivityThread.java:7319)
   at java.lang.reflect.Method.invoke(Native Method)
   at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
   at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:934)
```

Have same issue for regular retrofit without any retrofit-mock's

```
    @GET("url")
    suspend fun getClient(@Header("email") email: String): ClientPayload
```
   
```
    for method Api.getClient
        at retrofit2.Utils.methodError(Utils.java:52)
        at retrofit2.HttpServiceMethod.createCallAdapter(HttpServiceMethod.java:105)
        at retrofit2.HttpServiceMethod.parseAnnotations(HttpServiceMethod.java:66)
        at retrofit2.ServiceMethod.parseAnnotations(ServiceMethod.java:37)
        at retrofit2.Retrofit.loadServiceMethod(Retrofit.java:170)
        at retrofit2.Retrofit$1.invoke(Retrofit.java:149)
        at java.lang.reflect.Proxy.invoke(Proxy.java:397)
        at $Proxy0.getClient(Unknown Source)
        at com..android.data.ClientRegistrationRepository.getClient(ClientRegistrationRepository.kt:23)
        at com..android.domain.client.ClientNameUseCase$updateClientNameSettings$2.invokeSuspend(ClientNameUseCase.kt:21)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:241)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:740)
     Caused by: java.lang.IllegalArgumentException: Could not locate call adapter for class java.lang.Object.
      Tried:
       * retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
       * retrofit2.DefaultCallAdapterFactory
        at retrofit2.Retrofit.nextCallAdapter(Retrofit.java:241)
        at retrofit2.Retrofit.callAdapter(Retrofit.java:205)
        at retrofit2.HttpServiceMethod.createCallAdapter(HttpServiceMethod.java:103)
        	... 13 more
```

I am having the same problem as well without retrofit-mock.

```
    @GET("url")
    suspend fun getItem(): SomeResponse
```

```
   java.lang.IllegalArgumentException: Unable to create call adapter for class java.lang.Object
        for method SapService.getArticle
        at retrofit2.ServiceMethod$Builder.methodError(ServiceMethod.java:755)
        at retrofit2.ServiceMethod$Builder.createCallAdapter(ServiceMethod.java:240)
        at retrofit2.ServiceMethod$Builder.build(ServiceMethod.java:165)
        at retrofit2.Retrofit.loadServiceMethod(Retrofit.java:170)
        at retrofit2.Retrofit$1.invoke(Retrofit.java:147)
        at java.lang.reflect.Proxy.invoke(Proxy.java:1006)
        at $Proxy3.getArticle(Unknown Source)
        at ...
        at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:91)
        at kotlinx.coroutines.CoroutineScopeKt.coroutineScope(CoroutineScope.kt:186)
        at ...
        at ...
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.ResumeModeKt.resumeUninterceptedMode(ResumeMode.kt:45)
        at kotlinx.coroutines.internal.ScopeCoroutine.onCompletionInternal$kotlinx_coroutines_core(Scopes.kt:28)
        at kotlinx.coroutines.JobSupport.completeStateFinalization(JobSupport.kt:305)
        at kotlinx.coroutines.JobSupport.tryFinalizeFinishingState(JobSupport.kt:230)
        at kotlinx.coroutines.JobSupport.tryMakeCompleting(JobSupport.kt:799)
        at kotlinx.coroutines.JobSupport.makeCompletingOnce$kotlinx_coroutines_core(JobSupport.kt:742)
        at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:117)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
        at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:233)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
     Caused by: java.lang.IllegalArgumentException: Could not locate call adapter for class java.lang.Object.
      Tried:
       * com.jakewharton.retrofit2.adapter.kotlin.coroutines.CoroutineCallAdapterFactory
       * retrofit2.ExecutorCallAdapterFactory
        at retrofit2.Retrofit.nextCallAdapter(Retrofit.java:241)
        at retrofit2.Retrofit.callAdapter(Retrofit.java:205)
        at retrofit2.ServiceMethod$Builder.createCallAdapter(ServiceMethod.java:238)
        	... 28 more
```

--- a/retrofit-mock/pom.xml
+++ b/retrofit-mock/pom.xml
@@ -24,6 +24,11 @@
       <artifactId>jsr305</artifactId>
       <scope>provided</scope>
     </dependency>
+    <dependency>
+      <groupId>org.jetbrains.kotlin</groupId>
+      <artifactId>kotlin-stdlib</artifactId>
+      <optional>true</optional>
+    </dependency>
 
     <dependency>
       <groupId>junit</groupId>
@@ -35,10 +40,35 @@
       <artifactId>assertj-core</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.jetbrains.kotlinx</groupId>
+      <artifactId>kotlinx-coroutines-core</artifactId>
+      <scope>test</scope>
+    </dependency>
   </dependencies>
 
   <build>
     <plugins>
+      <plugin>
+        <groupId>org.jetbrains.kotlin</groupId>
+        <artifactId>kotlin-maven-plugin</artifactId>
+        <executions>
+          <execution>
+            <id>compile</id>
+            <phase>process-sources</phase>
+            <goals>
+              <goal>compile</goal>
+            </goals>
+          </execution>
+          <execution>
+            <id>test-compile</id>
+            <phase>test-compile</phase>
+            <goals>
+              <goal>test-compile</goal>
+            </goals>
+          </execution>
+        </executions>
+      </plugin>
       <plugin>
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-jar-plugin</artifactId>

--- a/retrofit-mock/src/main/java/retrofit2/mock/BehaviorDelegate.java
+++ b/retrofit-mock/src/main/java/retrofit2/mock/BehaviorDelegate.java
@@ -16,12 +16,18 @@
 package retrofit2.mock;
 
 import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
 import java.lang.reflect.Proxy;
 import java.lang.reflect.Type;
+import java.lang.reflect.WildcardType;
 import java.util.concurrent.ExecutorService;
 import javax.annotation.Nullable;
+import kotlin.coroutines.Continuation;
 import retrofit2.Call;
 import retrofit2.CallAdapter;
+import retrofit2.KotlinExtensions;
+import retrofit2.Response;
 import retrofit2.Retrofit;
 
 /**
@@ -54,11 +60,101 @@ public final class BehaviorDelegate<T> {
     final Call<R> behaviorCall = new BehaviorCall<>(behavior, executor, call);
     return (T) Proxy.newProxyInstance(service.getClassLoader(), new Class[] { service },
         (proxy, method, args) -> {
-          Type returnType = method.getGenericReturnType();
+          ServiceMethodAdapterInfo adapterInfo = parseServiceMethodAdapterInfo(method);
+
           Annotation[] methodAnnotations = method.getAnnotations();
           CallAdapter<R, T> callAdapter =
-              (CallAdapter<R, T>) retrofit.callAdapter(returnType, methodAnnotations);
-          return callAdapter.adapt(behaviorCall);
+              (CallAdapter<R, T>) retrofit.callAdapter(adapterInfo.responseType, methodAnnotations);
+
+          T adapted = callAdapter.adapt(behaviorCall);
+          if (!adapterInfo.isSuspend) {
+            return adapted;
+          }
+
+          Call<Object> adaptedCall = (Call<Object>) adapted;
+          Continuation<Object> continuation = (Continuation<Object>) args[args.length - 1];
+          try {
+            return adapterInfo.wantsResponse
+                ? KotlinExtensions.awaitResponse(adaptedCall, continuation)
+                : KotlinExtensions.await(adaptedCall, continuation);
+          } catch (Exception e) {
+            return KotlinExtensions.suspendAndThrow(e, continuation);
+          }
         });
+  }
+
+  /**
+   * Computes the adapter type of the method for lookup via {@link Retrofit#callAdapter} as well as
+   * information on whether the method is a {@code suspend fun}.
+   * <p>
+   * In the case of a Kotlin {@code suspend fun}, the last parameter type is a {@code Continuation}
+   * whose parameter carries the actual response type. In this case, we return {@code Call<T>} where
+   * {@code T} is the body type.
+   */
+  private static ServiceMethodAdapterInfo parseServiceMethodAdapterInfo(Method method) {
+    Type[] genericParameterTypes = method.getGenericParameterTypes();
+    if (genericParameterTypes.length != 0) {
+      Type lastParameterType = genericParameterTypes[genericParameterTypes.length - 1];
+      if (lastParameterType instanceof ParameterizedType) {
+        ParameterizedType parameterizedLastParameterType = (ParameterizedType) lastParameterType;
+        try {
+          if (parameterizedLastParameterType.getRawType() == Continuation.class) {
+            Type resultType = parameterizedLastParameterType.getActualTypeArguments()[0];
+            if (resultType instanceof WildcardType) {
+              resultType = ((WildcardType) resultType).getLowerBounds()[0];
+            }
+            if (resultType instanceof ParameterizedType) {
+              ParameterizedType parameterizedResultType = (ParameterizedType) resultType;
+              if (parameterizedResultType.getRawType() == Response.class) {
+                Type bodyType = parameterizedResultType.getActualTypeArguments()[0];
+                Type callType = new CallParameterizedTypeImpl(bodyType);
+                return new ServiceMethodAdapterInfo(true, true, callType);
+              }
+            }
+            Type callType = new CallParameterizedTypeImpl(resultType);
+            return new ServiceMethodAdapterInfo(true, false, callType);
+          }
+        } catch (NoClassDefFoundError ignored) {
+          // Not using coroutines.
+        }
+      }
+    }
+    return new ServiceMethodAdapterInfo(false, false, method.getGenericReturnType());
+  }
+
+  static final class CallParameterizedTypeImpl implements ParameterizedType {
+    private final Type bodyType;
+
+    CallParameterizedTypeImpl(Type bodyType) {
+      this.bodyType = bodyType;
+    }
+
+    @Override public Type[] getActualTypeArguments() {
+      return new Type[] { bodyType };
+    }
+
+    @Override public Type getRawType() {
+      return Call.class;
+    }
+
+    @Override public @Nullable Type getOwnerType() {
+      return null;
+    }
+  }
+
+  static class ServiceMethodAdapterInfo {
+    final boolean isSuspend;
+    /**
+     * Whether the suspend function return type was {@code Response<T>}.
+     * Only meaningful if {@link #isSuspend} is true.
+     */
+    final boolean wantsResponse;
+    final Type responseType;
+
+    ServiceMethodAdapterInfo(boolean isSuspend, boolean wantsResponse, Type responseType) {
+      this.isSuspend = isSuspend;
+      this.wantsResponse = wantsResponse;
+      this.responseType = responseType;
+    }
   }
 }

--- a/retrofit-mock/src/main/java/retrofit2/mock/KotlinExtensions.kt
+++ b/retrofit-mock/src/main/java/retrofit2/mock/KotlinExtensions.kt
@@ -1,0 +1,18 @@
+/*
+ * Copyright (C) 2019 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package retrofit2.mock
+
+inline fun <reified T> MockRetrofit.create(): BehaviorDelegate<T> = create(T::class.java)

--- a/retrofit-mock/src/test/java/retrofit2/mock/BehaviorDelegateKotlinTest.kt
+++ b/retrofit-mock/src/test/java/retrofit2/mock/BehaviorDelegateKotlinTest.kt
@@ -1,0 +1,166 @@
+/*
+ * Copyright (C) 2019 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package retrofit2.mock
+
+import kotlinx.coroutines.runBlocking
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Before
+import org.junit.Test
+import retrofit2.Response
+import retrofit2.Retrofit
+import java.io.IOException
+import java.util.Random
+import java.util.concurrent.TimeUnit.MILLISECONDS
+import java.util.concurrent.TimeUnit.NANOSECONDS
+
+class BehaviorDelegateKotlinTest {
+  internal interface DoWorkService {
+    suspend fun body(): String
+    suspend fun failure(): String
+    suspend fun response(): Response<String>
+    suspend fun responseWildcard(): Response<out String>
+  }
+
+  private val mockFailure = IOException("Timeout!")
+  private val behavior = NetworkBehavior.create(Random(2847))
+  private lateinit var service: DoWorkService
+
+  @Before fun before() {
+    val retrofit = Retrofit.Builder()
+        .baseUrl("http://example.com")
+        .build()
+    val mockRetrofit = MockRetrofit.Builder(retrofit)
+        .networkBehavior(behavior)
+        .build()
+    val delegate = mockRetrofit.create<DoWorkService>()
+
+    service = object : DoWorkService {
+      override suspend fun body(): String {
+        return delegate.returning(Calls.response("Response!")).body()
+      }
+
+      override suspend fun failure(): String {
+        val failure = Calls.failure<String>(mockFailure)
+        return delegate.returning(failure).failure()
+      }
+
+      override suspend fun response(): Response<String> {
+        val response = Calls.response("Response!")
+        return delegate.returning(response).response()
+      }
+
+      override suspend fun responseWildcard() = response()
+    }
+  }
+
+  @Test fun body() {
+    behavior.setDelay(100, MILLISECONDS)
+    behavior.setVariancePercent(0)
+    behavior.setFailurePercent(0)
+
+    val startNanos = System.nanoTime()
+    val result = runBlocking { service.body() }
+    val tookMs = NANOSECONDS.toMillis(System.nanoTime() - startNanos)
+
+    assertThat(tookMs).isGreaterThanOrEqualTo(100)
+    assertThat(result).isEqualTo("Response!")
+  }
+
+  @Test fun bodyFailure() {
+    behavior.setDelay(100, MILLISECONDS)
+    behavior.setVariancePercent(0)
+    behavior.setFailurePercent(100)
+
+    val startNanos = System.nanoTime()
+    val exception = runBlocking {
+      try {
+        throw AssertionError(service.body())
+      } catch (e: Exception) {
+        e
+      }
+    }
+    val tookMs = NANOSECONDS.toMillis(System.nanoTime() - startNanos)
+
+    assertThat(tookMs).isGreaterThanOrEqualTo(100)
+    assertThat(exception).isSameAs(behavior.failureException())
+  }
+
+  @Test fun failure() {
+    behavior.setDelay(100, MILLISECONDS)
+    behavior.setVariancePercent(0)
+    behavior.setFailurePercent(0)
+
+    val startNanos = System.nanoTime()
+    val exception = runBlocking {
+      try {
+        throw AssertionError(service.failure())
+      } catch (e: Exception) {
+        e
+      }
+    }
+    val tookMs = NANOSECONDS.toMillis(System.nanoTime() - startNanos)
+
+    assertThat(tookMs).isGreaterThanOrEqualTo(100)
+    // Coroutines break referential transparency on exceptions so compare type and message.
+    assertThat(exception).isExactlyInstanceOf(mockFailure.javaClass)
+    assertThat(exception).hasMessage(mockFailure.message)
+  }
+
+  @Test fun response() {
+    behavior.setDelay(100, MILLISECONDS)
+    behavior.setVariancePercent(0)
+    behavior.setFailurePercent(0)
+
+    val startNanos = System.nanoTime()
+    val result = runBlocking { service.response() }
+    val tookMs = NANOSECONDS.toMillis(System.nanoTime() - startNanos)
+
+    assertThat(tookMs).isGreaterThanOrEqualTo(100)
+    assertThat(result.body()).isEqualTo("Response!")
+  }
+
+  @Test fun responseFailure() {
+    behavior.setDelay(100, MILLISECONDS)
+    behavior.setVariancePercent(0)
+    behavior.setFailurePercent(100)
+
+    val startNanos = System.nanoTime()
+    val exception = runBlocking {
+      try {
+        throw AssertionError(service.response())
+      } catch (e: Exception) {
+        e
+      }
+    }
+    val tookMs = NANOSECONDS.toMillis(System.nanoTime() - startNanos)
+
+    assertThat(tookMs).isGreaterThanOrEqualTo(100)
+    assertThat(exception).isSameAs(behavior.failureException())
+  }
+
+  @Test fun responseWildcard() {
+    behavior.setDelay(100, MILLISECONDS)
+    behavior.setVariancePercent(0)
+    behavior.setFailurePercent(0)
+
+    val startNanos = System.nanoTime()
+    val result = runBlocking { service.responseWildcard() }
+    val tookMs = NANOSECONDS.toMillis(System.nanoTime() - startNanos)
+
+    assertThat(tookMs).isGreaterThanOrEqualTo(100)
+    assertThat(result.body()).isEqualTo("Response!")
+  }
+}

